### PR TITLE
vSphere cloud provider: re-use session for vCenter logins

### PR DESCRIPTION
**Pull request by dagnello**
This change allows for the re-use of a vCenter client session.  Addresses #34491

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34892)

### CODE REVIEW

**Comment by dagnello** [comment]
@mikedanese can we run tests on this

**Comment by mikedanese** [comment]
@k8s-bot ok to test

**Comment by pdhamdhere** [comment]
cc/ @abrarshivani Looks like good improvement. Can you please review?

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 1]
```diff
@@ -316,9 +333,57 @@ func vsphereLogin(cfg *VSphereConfig, ctx context.Context) (*govmomi.Client, err
...
+	}
+
+	m := session.NewManager(vs.client.Client)
+	// retrieve client's current session
+	u, err := m.UserSession(ctx)
+	if err != nil {
+		if soap.IsSoapFault(err) {
+			fault := soap.ToSoapFault(err).VimFault()
+			// If the PropertyCollector is not found, the saved session for this URL is not valid
+			if _, ok := fault.(types.ManagedObjectNotFound); ok {
```
Comment (pdhamdhere) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Do we care about this specific exception? Dont we want to create newClient if either `err != nil || u == nil` condition is true?
```
if err != nil || u != nil {
   return nil; // Success
}

// Log warning
vs.client.Logout(ctx)
vs.client, err = newClient(vs.cfg, ctx)

**Review by pdhamdhere — COMMENTED** [review]
Thanks for doing this. Can you please add "testing done" to your PR?

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 2]
```diff
@@ -1067,12 +1103,11 @@ func (vs *VSphere) DetachDisk(volPath string, nodeName k8stypes.NodeName) error
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	// Create vSphere client
-	c, err := vsphereLogin(vs.cfg, ctx)
+	// Ensure client is logged in and session is valid
+	err := vSphereLogin(vs, ctx)
 	if err != nil {
 		return err
```
Comment (abrarshivani) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: log error here

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 3]
```diff
@@ -1206,22 +1240,21 @@ func (vs *VSphere) DeleteVolume(vmDiskPath string) error {
...
```
Comment (abrarshivani) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: log error here

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 1 above)
Comment (dagnello) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: not really, will remove additional checks.

**Comment by dagnello** [comment]
Tested in vSphere 5.5 (vCenter 6.0).

**Review by pdhamdhere — APPROVED** [review]
LGTM.

**Pull request merged**